### PR TITLE
Add documentation section regarding AI assistance

### PR DESCRIPTION
**Pull request by dbutenhof**
## Summary

Assist in automated tracking of AI assistance usage by encouraging and standardizing on AI assistance tagging.

## Details

Red Hat has standardized on the recommended "Assisted-by" and "Generated-by" tagging convention to track the usage of AI coding assistants.

Propose some standardized representation.

## Test Plan

- [ ] Team review

## Related Issues

N/A

---

- [x] "I certify that all code in this PR is my own, except as noted below."

## Use of AI

- [ ] Includes AI-assisted code completion
- [ ] Includes code generated by an AI application
- [ ] Includes AI-generated tests (NOTE: AI written tests should have a docstring that includes `## WRITTEN BY AI ##`)

### CODE REVIEW

**Review by jaredoconnell — DISMISSED** [review]

File: DEVELOPING.md [hunk 1]
```diff
@@ -53,6 +53,43 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
+You can optionally add the LLM model name after the name of the AI coding assistant for clarity.
+
+For example:
+
+```markdown
+Assisted-by: Cursor
+Assisted-by: GitHub Copilot gpt-4o
+Assisted-by: Claude claude-3-5-sonnet-20240620
+Assisted-by: gemini-code-assist gemini-2.5-flash
+Assisted-by: openai-code-assist gpt-4o
```
Comment (sjmonson) [inline DEVELOPING.md]:
```suggestion
Assisted-by: GitHub Copilot <GPT-4.1>
Assisted-by: JetBrains AI <Gemini 3.1 Pro>
```

I would only give examples for tools that are not intended to do the full e2e generation here. Also wdyt about angle brackets for model name?

File: DEVELOPING.md [hunk 2]
```diff
@@ -53,6 +53,43 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
+
+```markdown
+Generated-by: <name of the AI coding assistant>
+```
+
+For example:
+
+```markdown
+Generated-by: Cursor
+Generated-by: GitHub Copilot gpt-4o
```
Comment (sjmonson) [inline DEVELOPING.md]:
```suggestion
Generated-by: Cursor <Claude 4.5 Sonnet>
Generated-by: claude-code <Claude 4.6 Sonnet>
Generated-by: Gemini Code Assist <Gemini 3 Flash>
```

Likewise to previous comment.

File: DEVELOPING.md [hunk 3]
```diff
@@ -53,6 +53,43 @@ The `-e` flag installs the package in editable mode, allowing you to make change
 
 3. **Update Documentation**: Update or add documentation to reflect your changes. This includes updating README files, docstrings, and any relevant guides.
 
+## Tag use of AI coding assistants
+
+When using AI coding assistants to generate any substantive code, you must add the following tag to the commit message:
```
Comment (sjmonson) [inline DEVELOPING.md]: Says you "must" here but line 91 says "optional" for certain cases. Maybe make it all one paragraph.

File: DEVELOPING.md [hunk 4]
```diff
@@ -53,6 +53,43 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
+Assisted-by: Cursor
+Assisted-by: GitHub Copilot gpt-4o
+Assisted-by: Claude claude-3-5-sonnet-20240620
+Assisted-by: gemini-code-assist gemini-2.5-flash
+Assisted-by: openai-code-assist gpt-4o
+```
+
+This does not apply to minor changes such as fixing typos, formatting, or small AI generated autocompletion suggestions that do not constitute substantive code.
+
+You may instead use the Generated-by tag to clarify that the commit was generated entirely by the AI coding assistant:
```
Comment (sjmonson) [inline DEVELOPING.md]: Mentioned this in our sync-up, but I think `Generated-by` should come before `Assisted-by` in the documentation since the majority of first-time contributors seems to be letting the AI handle most of the code/PR process rather then writing the code themselves which IMO falls under "Generated".

Also it probably makes sense to have one or two sentences in `AGENTS.md` on using `Generated-by`.

**Review by sjmonson — CHANGES_REQUESTED** [review]
Per our discussion in syncup. I think either put `Generated-by` first and

File: DEVELOPING.md (hunk 3 above)
Comment (dbutenhof) [reply DEVELOPING.md]: This is about commit messages; line 91 refers to code comments. It's a redirection and generalization of our less standard `## WRITTEN BY AI ##` marker -- but I didn't feel comfortable claiming it was mandatory. An "optional" requirement doesn't mean much in `AGENTS.md`, but this was intended more for human consumption.

File: DEVELOPING.md (hunk 1 above)
Comment (dbutenhof) [reply DEVELOPING.md]: I have mixed feelings about the model notation -- usage is very nonstandard. And while the current behavior of the P&D scanner shouldn't be considered definitive or even necessarily desirable, it will simply treat this whole thing as a unique permutation of "assistant" and break them all out.

I think for *code comments*, (depending on how introspective the model is), it should be possible to provoke the agent through `AGENTS.md` to insert the actual model it used to *create* a particular module or function; but at a *commit* level, after the fact, that may be difficult or impossible to determine ... and in fact may include multiple models.

With that said ... yeah, sure I pulled examples that include the model, and I do rather like using some delimiter. Whether any particular analysis tools will do anything rational with that is a whole different matter.

File: DEVELOPING.md (hunk 4 above)
Comment (dbutenhof) [reply DEVELOPING.md]: Yeah, I'll switch them around. You're right -- it makes sense to tackle the "whole cloth" model first and the "major edit" as a secondary case.

File: DEVELOPING.md (hunk 1 above)
Comment (sjmonson) [reply DEVELOPING.md]: Something I ran into right after leaving this comment is git parses trailers in that format as email addresses so maybe not angle brackets. But yeah, while I think it looks cleaner its not standard (not that any of this is really *standard*). So maybe just my other comment:


```suggestion
Assisted-by: GitHub Copilot GPT-4.1
Assisted-by: JetBrains AI Gemini 3.1 Pro
```

File: DEVELOPING.md [hunk 5]
```diff
@@ -53,6 +53,33 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
+- `Assisted-by: <name of the AI coding assistant>`: when the commit includes substantial code created or modified by the AI coding assistant.
+
+For example:
+
+```markdown
+Generated-by: Cursor
+Assisted-by: GitHub Copilot
+```
+
+You can optionally add the LLM model name after the name of the AI coding assistant for clarity, separated by a comma or space. (Do not enclose the model name in angle brackets, like `<gpt-4o>`, because GitHub will parse it as an email address.)
```
Comment (sjmonson) [inline DEVELOPING.md]:
```suggestion
You can optionally add the LLM model name after the name of the AI coding assistant for clarity, separated by a comma or space.
```

I don't think we need to state this part since I don't think anyone is doing it except me.

File: DEVELOPING.md [hunk 6]
```diff
@@ -53,6 +53,33 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
+
+For example:
+
+```markdown
+Generated-by: Cursor claude-4.6-sonnet
+Assisted-by: GitHub Copilot gpt-4o
+```
+
+This rule does not apply to minor changes such as fixing typos, formatting, or isolated AI generated syntax suggestions that do not constitute substantive logic.
+
```
Comment (sjmonson) [inline DEVELOPING.md]:
```suggestion
```

I am thinking omit this since we would prefer all AI contributions are labeled. Above already says "commit was generated primarily" and "commit includes substantial code created or modified" which kinda says this implicitly.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -33,6 +33,7 @@ Link any relevant issues that this PR addresses.
 
 ## Use of AI
 
-- [ ] Includes AI-assisted code completion
-- [ ] Includes code generated by an AI application
-- [ ] Includes AI-generated tests (NOTE: AI written tests should have a docstring that includes `## WRITTEN BY AI ##`)
+- [ ] Includes code generated or substantially modified by an AI agent
+- [ ] Includes tests generated or substantially modified by an AI agent
+
+NOTE: the `Generated-by` and `Assisted-by` tags should be used in both git commit messages and docstrings to indicate when code or tests were generated or substantially modified by an AI agent, as described in the project's [`DEVELOPING.md`](https://github.com/vllm-project/guidellm/blob/main/DEVELOPING.md) file.
```
Comment (sjmonson) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
> NOTE: the `Generated-by` and `Assisted-by` tags should be used in both git commit messages and docstrings to indicate when code or tests were generated or substantially modified by an AI agent, as described in the project's [`DEVELOPING.md`](https://github.com/vllm-project/guidellm/blob/main/DEVELOPING.md) file.
```

**Review by sjmonson — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 7 above)
Comment (dbutenhof) [reply .github/PULL_REQUEST_TEMPLATE.md]: I was tempted to use `> [!NOTE]\` or even `> [!IMPORTANT]\` ... except that while the template is markdown and it'd presumably render, in the end this becomes a git commit description which won't do any formatting. But the `>` won't obfuscate the commit description and does set off the note a bit in the template -- so, sure.

File: DEVELOPING.md (hunk 5 above)
Comment (dbutenhof) [reply DEVELOPING.md]: I debated that for whole seconds; but you're probably right that it's excessive noise that might make contributors less likely to read the whole thing. 😆

File: DEVELOPING.md (hunk 6 above)
Comment (dbutenhof) [reply DEVELOPING.md]: Fair enough. I considered the extra warning potentially useful just because the minor stuff isn't interesting to track -- and when we reach the point where every commit uses some AI we lose all resolution.

But, again, it *is* implied elsewhere and maybe the "less noise is better" rule should apply until/unless we find it becomes an issue.

File: DEVELOPING.md [hunk 8]
```diff
@@ -55,40 +55,28 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
 
 ```markdown
-Generated-by: Cursor
-Generated-by: GitHub Copilot gpt-4o
+Generated-by: Cursor claude-4.6-sonnet
+Assisted-by: GitHub Copilot gpt-4o
 ```
 
-Use of the Assisted-by or Generated-by tags in code (comments, docstrings, etc.), where an AI assistant has generated or substantially modified a function, class, or module, is optional but encouraged for clarity and transparency.
+Use of the Generated-by and Assisted-by tags in code (comments, docstrings, etc.) is strongly encouraged, to help identify where an AI assistant has generated or substantially modified a function, class, or module. GuideLLM has also recognized the tag `## WRITTEN BY AI ##` in a docstring or comment to indicate that test cases were written by an AI assistant, but `Generated-by <assistant>` or `Assisted-by <assistant>` are preferred for clarity and consistency.
```
Comment (jaredoconnell) [inline DEVELOPING.md]: I think we should be more clear about the difference in role between these. The test docstring is so developers can get an idea of how seriously to value a test. The commit tags are more for legal and statistical reasons.

Comment (dbutenhof) [reply DEVELOPING.md]: I don't think there's any difference we'd want to (or should) explain at this level. That is, we're not going to say "if you admit to using an AI agent to generate tests, we'll view the changes with greater suspicion".

If anything, maybe we could establish a general background with something like "We are soliciting your assistance in this as we have an organizational mandate to track and quantify the usage of AI in our code base." But that's somewhat implied, and doesn't really sound like it's adding much context.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 7 above)
Comment (sjmonson) [reply .github/PULL_REQUEST_TEMPLATE.md]: It also does not work in PRs. GitHub only supports that format for issues and in-repo docs.

File: DEVELOPING.md (hunk 8 above)
Comment (sjmonson) [reply DEVELOPING.md]: I misunderstood this on the first pass. I don't think we want to encourage inline Assisted-by / Generated-by tags in code. That will just get messy. We make an exception for `## WRITTEN BY AI ##` because its very constrained to a very specific location within each test. Its also not about "suspicion" its about determining how intentional a test is, is it just testing made up edge cases or did a human specifically determine that a gap in testing.

Tldr. my preference would be to drop this paragraph.

Comment (dbutenhof) [reply DEVELOPING.md]: Hmm. Yes, putting them everywhere could become a pain -- although in principle the idea seemed appealing, I'll grant that in practice it'd probably be less so. However, if we're even going to tag AI tests, I'd still prefer to use Generated-by rather than the odd ## WRITTEN BY AI ##.

File: DEVELOPING.md [hunk 9]
```diff
@@ -53,6 +53,35 @@ The `-e` flag installs the package in editable mode, allowing you to make change
 
 3. **Update Documentation**: Update or add documentation to reflect your changes. This includes updating README files, docstrings, and any relevant guides.
 
+## Tag use of AI coding assistants
+
+When AI coding assistants are used to generate or modify substantive code, the GuideLLM project requires that you add one of the following two "trailer" tags to the git commit message. Git "trailers" are a series of key-value pairs appended to the end of the commit message, separated by a line break, including `Co-authored-by: <name> <email>` and `Signed-off-by: <name> <email>`. We use the keys `Generated-by` and `Assisted-by` to indicate when code or tests were generated or substantially modified by an AI agent.
```
Comment (sjmonson) [inline DEVELOPING.md]:
```suggestion
When AI coding assistants are used to generate or modify substantive code, the GuideLLM project requires that you add one of the following trailers to the git commit message:
```

Keep it simple

File: DEVELOPING.md [hunk 10]
```diff
@@ -53,6 +53,35 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
+Signed-off-by: John Doe <john.doe@example.com>
+```
+
+You can optionally add the LLM model name after the name of the AI coding assistant, separated by a comma or space.
+
+For example:
+
+```markdown
+Generated-by: Cursor claude-4.6-sonnet
+Assisted-by: GitHub Copilot, gpt-4o
```
Comment (sjmonson) [inline DEVELOPING.md]:
```suggestion
Assisted-by: GitHub Copilot gpt-4o
```

We decided against brackets, I don't think we should add a comma.

File: DEVELOPING.md [hunk 11]
```diff
@@ -53,6 +53,35 @@ The `-e` flag installs the package in editable mode, allowing you to make change
...
+```
+
+Do not use the `Generated-by` and `Assisted-by` tags in code (comments, docstrings, etc.): as the code evolves, this sort of "origin" attribution can lose relevance and become noise.
```
Comment (sjmonson) [inline DEVELOPING.md]:
```suggestion
```

Lets not add "don't do this" until we actually have a problem with people doing it. Plus if you want to switch from `## WRITTEN BY AI ##` to `Generated-by:` in tests, this directly contradicts that.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 12]
```diff
@@ -33,6 +33,7 @@ Link any relevant issues that this PR addresses.
 
 ## Use of AI
 
-- [ ] Includes AI-assisted code completion
-- [ ] Includes code generated by an AI application
-- [ ] Includes AI-generated tests (NOTE: AI written tests should have a docstring that includes `## WRITTEN BY AI ##`)
+- [ ] Includes code generated or substantially modified by an AI agent
+- [ ] Includes tests generated or substantially modified by an AI agent
+
+> NOTE: the `Generated-by` and `Assisted-by` tags should be used in both git commit messages and docstrings to indicate when code or tests were generated or substantially modified by an AI agent, as described in the project's [`DEVELOPING.md`](https://github.com/vllm-project/guidellm/blob/main/DEVELOPING.md) file.
```
Comment (sjmonson) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
> NOTE: the `Generated-by` and `Assisted-by` tags should be used in git commit messages to indicate when code is generated or substantially modified by an AI agent, as described in the project's [`DEVELOPING.md`](https://github.com/vllm-project/guidellm/blob/main/DEVELOPING.md) file.
```

Keep it simple

**Review by sjmonson — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 12 above)
Comment (dbutenhof) [reply .github/PULL_REQUEST_TEMPLATE.md]: Dang -- I forgot to update the template file.

File: DEVELOPING.md (hunk 10 above)
Comment (dbutenhof) [reply DEVELOPING.md]: I was motivated by the fact that our existing attributions include commas. However, yeah: the fact that some exist doesn't mean we need to encourage that style in the future.

**Review by sjmonson — APPROVED** [review]

**Review by jaredoconnell — APPROVED** [review]
